prime-router/src/main/kotlin/azure/ActionHistory.kt

```diff
@@ -713,4 +732,166 @@ class ActionHistory {
             }
         }
     }
+
+    /**
+     * Creates a string that will be used to populate the action_response column of an action
+     * after that action has been completed
+     * @param options Message options passed in
+     * @param warnings Store of warnings generated during the pipeline
+     * @param errors Store of errors generated during the pipeline
+     * @param verbose If true, all item routing details will be included in the response
+     * @param actionHistory The ongoing action history instance being used
+     * @param report The report this response body is for
+     */
+    fun createResponseBody(
+        options: Options,
+        warnings: List<ResultDetail>,
+        errors: List<ResultDetail>,
+        verbose: Boolean,
+        actionHistory: ActionHistory? = null,
+        report: Report? = null
+    ): String {
+        val factory = JsonFactory()
+        val outStream = ByteArrayOutputStream()
+        factory.createGenerator(outStream).use {
+            it.useDefaultPrettyPrinter()
+            it.writeStartObject()
+            if (report != null) {
+                it.writeStringField("id", report.id.toString())
+                it.writeStringField("timestamp", DateTimeFormatter.ISO_INSTANT.format(Instant.now()))
+                it.writeStringField("topic", report.schema.topic)
+                it.writeNumberField("reportItemCount", report.itemCount)
+                // TODO: need to get the correct path and set the ENDPOINT_BASE correctly, blocked by waiting for
+                //  the historyApi to be ready.
+                // it.writeStringField("location", "[base path]/${report.id}")
+            } else
+                it.writeNullField("id")
+
+            actionHistory?.prettyPrintDestinationsJson(it, WorkflowEngine.settings, options)
+            // print the report routing when in verbose mode
+            if (verbose) {
+                it.writeArrayFieldStart("routing")
+                createItemRouting(report, actionHistory).forEach { ij ->
+                    it.writeStartObject()
+                    it.writeNumberField("reportIndex", ij.reportIndex)
+                    it.writeStringField("trackingId", ij.trackingId)
+                    it.writeArrayFieldStart("destinations")
+                    ij.destinations.sorted().forEach { d -> it.writeString(d) }
+                    it.writeEndArray()
+                    it.writeEndObject()
+                }
+                it.writeEndArray()
+            }
+
+            it.writeNumberField("warningCount", warnings.size)
+            it.writeNumberField("errorCount", errors.size)
+
+            fun writeDetailsArray(field: String, array: List<ResultDetail>) {
```

Comment:
kdoc

Reply:
I will suggest that you move all this functions that are inside this other function and write unit tests for them.

Reply:
Ticketed: #2945